common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java

```diff
@@ -198,181 +177,463 @@
+            try {
+                loadAllExtensionClass(loader);
+                return getExtension(defaultExtensionURL, loader, argTypes, args);
+            } catch (Throwable e) {
+                if (e instanceof EnhancedServiceNotFoundException) {
+                    throw (EnhancedServiceNotFoundException)e;
+                } else {
+                    throw new EnhancedServiceNotFoundException(
+                            "not found service provider for : " + type.getName() + " caused by " + ExceptionUtils
+                                    .getFullStackTrace(e));
+                }
             }
+        }
 
-            if (extensions.isEmpty()) {
-                throw new EnhancedServiceNotFoundException(
-                    "not found service provider for : " + service.getName() + "[" + activateName
-                        + "] and classloader : " + ObjectUtils.toString(loader));
+        @SuppressWarnings("rawtypes")
+        private S loadExtension(String activateName, ClassLoader loader, Class[] argTypes,
+                                Object[] args) {
+            if (io.seata.common.util.StringUtils.isEmpty(activateName)) {
+                throw new IllegalArgumentException("the name of service provider for [" + type.getName() + "] name is null");
+            }
+            try {
+                loadAllExtensionClass(loader);
+                ExtensionURL cachedExtensionURL = getCachedServiceURL(activateName);
+                return getExtension(cachedExtensionURL, loader, argTypes, args);
+            } catch (Throwable e) {
+                if (e instanceof EnhancedServiceNotFoundException) {
+                    throw (EnhancedServiceNotFoundException)e;
+                } else {
+                    throw new EnhancedServiceNotFoundException(
+                            "not found service provider for : " + type.getName() + " caused by " + ExceptionUtils
+                                    .getFullStackTrace(e));
+                }
             }
-            Class<?> extension = extensions.get(extensions.size() - 1);
-            S result = initInstance(service, extension, argTypes, args);
-            if (!foundFromCache && LOGGER.isInfoEnabled()) {
-                LOGGER.info("load " + service.getSimpleName() + "[" + activateName + "] extension by class[" + extension
-                    .getName() + "]");
+        }
+
+        private S getExtension(ExtensionURL url, ClassLoader loader, Class[] argTypes,
+                               Object[] args) {
+            if (url == null) {
+                throw new EnhancedServiceNotFoundException("service provider not found");
             }
-            return result;
-        } catch (Throwable e) {
-            if (e instanceof EnhancedServiceNotFoundException) {
-                throw (EnhancedServiceNotFoundException)e;
+            if (Scope.SINGLETON.equals(url.getScope())) {
+                Holder<Object> holder = extensionInstances.get(url);
```

Comment:
What's the benefit to use Holder here? Is Holder class necessary?

Reply:
This part of the code which use Holder and Double-Checkd Locking refers to Dubbo SPI. I think it has two functions at present:
1. When concurrent, ensure that an extensionDefinition use the same lock object
2. Whether the value in the Holder is null can be used as the judgment condition of Double-Checked Locking
Without using Holder here,i think  it is difficult to guarantee the above two points at the same time.

Reply:
This part of the code which use Holder and Double-Checkd Locking refers to Dubbo SPI. I think it has two functions at present:
1.When concurrent, ensure that an extensionDefinition use the same lock object
2.Whether the value in the Holder is null can be used as the judgment condition of Double-Checked Locking
Without using Holder here,i think it is difficult to guarantee the above two points at the same time.